Add natspec comments parser

obernardovieira opened the pull request:
This is related to https://github.com/solidityj/solidity-antlr4/pull/19
and tries to solve #10 and #20 
I'm using some code from #20

obernardovieira commented:
Does it really worth to organize the comments? Like, type, text?
I'm having difficulties with ANTLR.

federicobond commented:
Good work @obernardovieira! Yes, please upload changes in lib directory. Also update the tests to actually assert the shape of the parsed comments. Each NatSpec tag should be parsed separately so it can be easily consumed.

obernardovieira commented:
"Each NatSpec tag should be parsed to a separate object key so it can be easily consumed."

You mean, something like 
```json
{ 
      "type": "ContractDefinition",
      "natSpec": {
            "dev": "some dev message",
            "param": [
                  {
                        "name": "user",
                        "type": "address",
                        "message": "the user address"
                  }
            ]
      },
      "name": "Sum",
      "baseContracts": [],
      "subNodes": [],
      "kind": "contract",
}
```

or, can it be a simple string, like `/** @dev hello\n* @param stuff some\n*/`?

federicobond commented:
Yes, I meant a separate object key, so it can be easily consumed by others.

obernardovieira commented:
Since I don't know how to prase it with antlr, will it be fair to do some parser in JS code? Or, can you give me a hand, please?

federicobond commented:
Yes, it would be actually be better to do it in JS in this case. Add a function that takes the content of the natspec comment and returns its object representation. It can live in a separate file.

obernardovieira commented:
Hello,
I've turned it into an object and I'm now testing it correctly.
Please, give me some feedback when you can.
Might be cool to have some opinion from @chmanie @onbjerg and @0Alic as well.

federicobond commented:
This is looking really good @obernardovieira! 🙌 

I would most of the code from the `NatSpec` function to a separate file. There should be a function (we can call it `parseNatSpec`) that takes a string with the full comment and returns an object. Consider splitting that code into functions too or adding comments to make it easier to understand.

obernardovieira commented:
Thanks, appreciate :raised_hands: 
Moving it to a file was a good idea.
I don't think it worths to break into methods, It is a simple one.
Give me your opinion and if cool, let's merge it. :tada:

federicobond commented:
I'd remove the `isMultiline` from the NatSpec node, as it's a detail that belongs to the concrete syntax, not AST.

I would also remove the `type` key and put the contents of `comments` top-level, making natspec no longer an AST node by itself, but rather a component of its parent node.

Finally, I would format the params as an object with parameter names as keys, to be more aligned to the official style: https://solidity.readthedocs.io/en/develop/natspec-format.html#developer-documentation

federicobond commented on src/ASTBuilder.js:
```diff
@@ -1,4 +1,5 @@
 const antlr4 = require('./antlr4/index')
+const ParseComments = require('./ParseNatSpec');
```
Rename this file `natspec.js` and lowercase the parseComments function name please.

federicobond commented on src/ParseNatSpec.js:
```diff
@@ -0,0 +1,75 @@
+  for (let x = 1; x < splitComments.length; x += 2) {
+
+    // verify if the type exists in the map
+    let previousValue = resultComments.get(splitComments[x]);
```
For clarity, please extract `splitComments[x]` to a variable named `key` or similar, and extract `splitComments[x + 1]` to `value` or similar.

obernardovieira replied:
great, done

federicobond commented on src/ParseNatSpec.js:
```diff
@@ -0,0 +1,75 @@
+  const splitComments = comment
+    .split(/@(title|author|notice|dev|param|return)/g);
+  // let's start a map
+  const resultComments = new Map();
```
Why not use a simple object literal here?

obernardovieira replied:
right, much better. Thank you

federicobond commented on src/ParseNatSpec.js:
```diff
@@ -0,0 +1,75 @@
+    }
+  }
+  // convert map to pseudo-json
+  const obj = Object.create(null);
```
Create the object using a simple literal: `{}`

obernardovieira replied:
I removed it. Was not needed

obernardovieira commented:
what do you mean by "I would format the params as an object with parameter names as keys"
```javascript
param: [
    { stuff: "some food" },
    { no: "more pizza today" },
    { yes: "please, another one tomorrow" }
]
```
isn't this correct?
BTW, I've realized I still miss a few things. Thanks.

federicobond commented:
I mean more like the format used here: https://solidity.readthedocs.io/en/develop/natspec-format.html#developer-documentation where the name of the parameter is the object key, as opposed to an array of params.

obernardovieira commented:
I'm confused. I updated the comments with something more realistic (they were a bit stupid, sorry about that) and I've added some new tests.
What I see is
```javascript
{
  "author" : "Larry A. Gardner",
  "details" : "All function calls are currently implemented without side effects",
  "methods" :
  {
    "age(uint256)" :
    {
      "author" : "Mary A. Botanist",
      "details" : "The Alexandr N. Tetearing algorithm could increase precision",
      "params" :
      {
        "rings" : "The number of rings from dendrochronological sample"
      },
      "return" : "age in years, rounded up for partial years"
    }
  },
  "title" : "A simulator for trees"
}
```
And what I have is similar, like, a function definition node, will have a natspec field containing
```javascript
dev: 'This method transfer fund to other user',
param:[
    { from: 'the address extract funds' },
    { to: 'the user address to give funds' },
    { amount: 'the amount to transfer' },
],
```

federicobond commented:
Is it OK if I push some code to your branch? It will be easier to show in code.

obernardovieira commented:
Yes, of course.

federicobond commented:
There, take a look a this commit: https://github.com/federicobond/solidity-parser-antlr/pull/71/commits/ff159869db66d84cd8c766a6562bbdce7bb23c02

obernardovieira commented:
ahh yeah, very obvious :man_facepalming: sorry about that :joy: 
Do you think it's good to merge?

obernardovieira commented:
HI, I just remember, actually solidityj/solidity-antlr4#19 needs to be merged before this one, but I also have one question that I forgot to ask before. Is it possible to extract comment from global variables? I've tried, but without success.

obernardovieira commented:
@federicobond any feedback?

federicobond commented:
Sorry @obernardovieira, I have been pretty busy at work. I will look into this in the weekend.

federicobond commented:
Merged manually https://github.com/federicobond/solidity-parser-antlr/commit/2273972a655e10cf53c8ac8a9ded5c1f9e87fca6. Thank you!

obernardovieira commented:
Cool, thank you.
Could you please close #10 and #20?
Thank you :pray:
